### ISSUE TITLE
feat: add MarkupString support in FluentDataGrid

### ISSUE THREAD

**Issue by thebarrettlo** [microsoft/fluentui-blazor#1764]
# 🙋 Feature Request

Add support for [`MarkupString`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.markupstring?view=aspnetcore-8.0)s to be rendered within `FluentDataGrid` cells when the content is populated through `PropertyColumn`s.

## 🤔 Expected Behavior

`MarkupString` objects can be used in `FluentDataGrid`s so that additional HTML can be dynamically rendered within cells.

## 😯 Current Behavior

Currently, trying to pass in `MarkupString` objects with HTML tags results in the HTML being rendered as plaintext.

## 💁 Possible Solution

It looks like the `PropertyColumn` currently passes a string to the render tree builder that is created from the `_cellTextFunc`. This function is defined to return a string, so even setting the `TProp` type param as `MarkupString` does not do anything.

One potential solution could be a switch parameter added to `PropertyColumn`, and if `true`, use a different cell text function that outputs a `MarkupString`. Another solution could be to add special logic that checks if `TProp` is `MarkupString`, and if so, do the same.

## 🔦 Context

I'm querying data that has text occasionally containing links (i.e., "https://..."). This data is displayed in a `FluentDataGrid`, but as it is now, the links are plaintext and are not hyperlinked. I have logic in place to identify these snippets and wrap them with HTML `a` tags and cast them to `MarkupString`s, but `PropertyColumn` converts these back to `string`s before any of the markup parsing can take place.

## 💻 Examples

```html
<fluent-data-grid-cell cell-id="c10" cell-type="default" grid-column="4" class="col-justify-start multiline-text" title="..." aria-label="..." b-w6qdxfylwy="" tabindex="-1" role="gridcell" style="grid-column: 4;">
  #shadow-root (open)
  "Most Azure services and client SDKs include a retry mechanism. However, these differ because each service has different characteristics and requirements, and so each retry mechanism is tuned to a specific service

  Reference: <a href="https://learn.microsoft.com/en-gb/azure/architecture/best-practices/retry-service-specific#azure-storage-retry-guidelines" target="_blank">https://learn.microsoft.com/en-gb/azure/architecture/best-practices/retry-service-specific#azure-storage-retry-guidelines</a>"
  <!---->
</fluent-data-grid-cell>
```

![image](https://github.com/microsoft/fluentui-blazor/assets/36118918/e1459b9d-b9ff-483c-ac0a-8d121b5c69b9)

**Comment by vnbaaij** [comment]
What if I just change the `CellContent` method to
```
protected internal override void CellContent(RenderTreeBuilder builder, TGridItem item)
=> builder.AddContent(0, new MarkupString(_cellTextFunc?.Invoke(item) ?? string.Empty));
```

Gives me this when cpoying your example in the multiline text content example:
![image](https://github.com/microsoft/fluentui-blazor/assets/1761079/ac2e8e78-cc3a-4b85-8a0e-9879c26479a3)

Don't see any adverse effects on any of the other examples.

Thoughts?

**Comment by thebarrettlo** [comment]
Looks like that would work well! I think there should be a way to disable it though (or have it disabled by default) in case that the cell content is being populated by data from an untrusted source and it happens to have malicious HTML. Especially since the official Blazor docs note this when using `MarkupString`: https://learn.microsoft.com/en-us/aspnet/core/blazor/components/?view=aspnetcore-8.0#raw-html:~:text=Rendering%20raw%20HTML%20constructed%20from%20any%20untrusted%20source%20is%20a%20security%20risk%20and%20should%20always%20be%20avoided.

**Comment by vnbaaij** [comment]
Agree. Was looking at that too. Maybe add an overloaded method so we don't break existing implementations

**Comment by vnbaaij** [comment]
I'm implementing it like this:

```
 protected internal override void CellContent(RenderTreeBuilder builder, TGridItem item)
 => builder.AddContent(0, _cellTextFunc?.Invoke(item));

 protected internal override void CellContent(RenderTreeBuilder builder, TGridItem item, bool allowRawHtml)
 => builder.AddContent(0, new MarkupString(_cellTextFunc?.Invoke(item) ?? string.Empty));
```
So a new overloaded method and a new `bool AllowRawHtml` parameter.

When using this new variant on a `TemplateColumn` it throws a `NotImplementedException` as it is not possible to pass the parameter to the ChildContent it renders.

**Comment by dvoituron** [comment]
Why not use a `TemplateColumn` with `MarkupString`, if you have HTML content? 

Example.
```xml
<FluentDataGrid Items="@people">
    <PropertyColumn Property="@(p => p.PersonId)" Sortable="true" />
    <PropertyColumn Property="@(p => p.Name)" Sortable="true" />
    <TemplateColumn Sortable="true" Class="multiline-text">
        @((MarkupString)context.Description)
    </TemplateColumn>
</FluentDataGrid>
```

**Comment by thebarrettlo** [comment]
@dvoituron thanks for the suggestion, just tested it with my use case and that does work. I think there would be value in having an "allow HTML" parameter to `PropertyColumn` like Vincent suggested because it's not obvious (to me) that the property is always converted to a `string` thanks to the `_cellTextFunc` (even a `MarkupString`).

But maybe the use case of wanting to parse/render raw HTML inside a `DataGrid` is too rare to warrant a new feature like that. My use case is solved, so I'll be happy either way.

**Comment by thebarrettlo** [comment]
Oh, I'll add that the built-in sorting of `PropertyColumn` is convenient :)

**Comment by vnbaaij** [comment]
We'll keep it as-is. `TemplateColumn` solves the issue. If some extra work is then needed for sorting, then so be it.
